### ISSUE TITLE
state of the art performance?

### ISSUE THREAD

**Issue by eric-haibin-lin** [rdspring1/PyTorch_LM#2]
Nice work! I have a question regarding the result:
In the paper "Exploring the limits of language modeling", it reports test ppl of 54.1 using LSTM-512-512. Does it mean two 2 layers are used in the paper, while your result is obtained from 4 layers ? If so, what makes the difference?

**Comment by rdspring1** [comment]
Sorry for the slow reply! I've been busy fine-tuning the hyperparameters, fixing small bugs, and reproducing the results. The latest model reports test ppl of 46.47 using LSTM 2048-256 - a single-layer lstm with 256 projection layer. You can run the model on a single GPU - (1080 Ti or higher) to reproduce the results.

**Comment by eric-haibin-lin** [comment]
Thanks for the reply! As I am reading the reference "Exploring the Limits of Language Modeling"
 and your code, I noticed that the original "LSTM Projected" cells actually performs the projection and pass the projected result [as the state to the next time step](https://github.com/rafaljozefowicz/lm/blob/master/model_utils.py#L107-L110). But in your implementation, you're projecting only the LSTM result, not passing it as states between time steps. I am wondering if that intended? Please correct me if I misunderstood anything. Thank you very much for you kind attention.

**Comment by rdspring1** [comment]
Yes, it was intended. I wanted to use the pyTorch LSTM module with its cuDNN optimizations. It currently doesn't support the projection operation on the internal state. I also believed the projection layer is primarily used to reduce the number of parameters in the embedding and softmax layers. 

Passing the projected result as the state for the next step should provide a minor speed boost.

**Comment by eric-haibin-lin** [comment]
I see. That makes sense. 

Also I noticed that when calculating the loss, tensorflow implementation uses a mask which sets 0 on sentence boundaries, while yours doesn't. Is that also intentional? That would give different result on evaluation, right?

**Comment by rdspring1** [comment]
Where is that in the TensorFlow / PyTorch implementation?

**Comment by eric-haibin-lin** [comment]
Oh, it's the TF github mentioned in your references in README. 
https://github.com/rafaljozefowicz/lm/blob/master/language_model.py#L102-L103
https://github.com/rafaljozefowicz/lm/blob/master/data_utils.py#L100-L110

**Comment by rdspring1** [comment]
I don't believe it should make a difference. The weights are always set to 1, since the entire vector is filled.

This repository removed the weight vector completely. (ICLR 2017 - Factorized LSTM) 
It was forked from the original repo. [https://github.com/okuchaiev/f-lm](https://github.com/okuchaiev/f-lm)

**Comment by eric-haibin-lin** [comment]
Right, I misinterpreted it. It's always filled with ones except that there could be a few zeros for the last batch. Shouldn't make a big difference.
